src/Features/Core/Portable/Testing/AbstractTestMethodFinder.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Features.Testing;
+
+internal abstract class AbstractTestMethodFinder<TMethodDeclaration>(IEnumerable<ITestFrameworkMetadata> testFrameworks) : ITestMethodFinder where TMethodDeclaration : SyntaxNode
+{
+    /// <summary>
+    /// Output the method symbol as a fully qualified method name, e.g. Namespace.Class.Method to match what test discovery gives us.
+    /// Generics are not applicable here - none of our supported test frameworks allow generic test classes / methods.
+    /// </summary>
+    private static readonly SymbolDisplayFormat s_methodSymbolNoParametersDisplayFormat = new(
+        typeQualificationStyle: SymbolDisplayTypeQualificationStyle.NameAndContainingTypesAndNamespaces,
+        memberOptions: SymbolDisplayMemberOptions.IncludeContainingType);
+
+    /// <summary>
+    /// Sometimes the test runner outputs fully qualified method names with parameters.  So we also check for that scenario.
+    /// </summary>
+    private static readonly SymbolDisplayFormat s_methodSymbolWithParametersDisplayFormat = new(
+        typeQualificationStyle: SymbolDisplayTypeQualificationStyle.NameAndContainingTypesAndNamespaces,
+        memberOptions: SymbolDisplayMemberOptions.IncludeContainingType | SymbolDisplayMemberOptions.IncludeParameters);
+
+    protected readonly ImmutableArray<ITestFrameworkMetadata> TestFrameworkMetadata = testFrameworks.ToImmutableArray();
+
+    protected abstract Task<bool> IsTestMethodAsync(Document document, TMethodDeclaration method, CancellationToken cancellationToken);
+
+    public async Task<ImmutableArray<SyntaxNode>> GetPotentialTestMethodsAsync(TextSpan textSpan, Document document, CancellationToken cancellationToken)
```

Comment:
would prefer this be `Document` then `TextSpan` to match our patterns of "Wider, than narrower" on method inputs.